src/gt4py/next/iterator/transforms/trace_shifts.py

```diff
@@ -322,46 +327,48 @@ def fun(*args):
 
         return fun
 
-    def visit_StencilClosure(self, node: ir.StencilClosure):
-        tracers = []
-        for inp in node.inputs:
+    def initialize_context(self, inputs: Iterable[ir.Sym | ir.SymRef]) -> dict[str, Any]:
+        ctx: dict[str, Any] = {**_START_CTX}
+        for inp in inputs:
             self.shift_recorder.register_node(inp)
-            tracers.append(IteratorArgTracer(arg=inp, shift_recorder=self.shift_recorder))
-
-        result = self.visit(node.stencil, ctx=_START_CTX)(*tracers)
-        assert all(el is Sentinel.VALUE for el in _primitive_constituents(result))
-        return node
+            ctx[inp.id] = ArgTracer(arg=inp, shift_recorder=self.shift_recorder)
+        return ctx
 
     @classmethod
-    def apply(
-        cls, node: ir.StencilClosure | ir.FencilDefinition, *, inputs_only=True, save_to_annex=False
-    ) -> (
-        dict[int, set[tuple[ir.OffsetLiteral, ...]]] | dict[str, set[tuple[ir.OffsetLiteral, ...]]]
+    def trace_stencil(
+        cls, stencil: ir.Expr, *, num_args: Optional[int] = None, save_to_annex: bool = False
     ):
+        if isinstance(stencil, ir.Lambda):
+            assert num_args is None or num_args == len(stencil.params)
+            num_args = len(stencil.params)
```

Comment:
what's special with lambdas? what else can it be?

Reply:
For lambdas we can deduce the number of arguments and the `num_args` parameter is optional. This is essentially a convenience feature for testing. I've added a comment.